moto/s3/models.py

```diff
@@ -339,6 +339,16 @@ def is_locked(self):
 
         return False
 
+    def dispose(self):
+        if self.multipart is not None:
+            for part in self.multipart.parts.values():
+                part.dispose()
+        if self._value_buffer:
+            self._value_buffer.close()
+
+    def __del__(self):
+        self.dispose()
+
```

Comment:
I think it is a good idea to call `dispose()` in `__del__()` as a last resort. In the future, there may be additional clean-up for other per-key resources. It's worth noting that this is equivalent to what the Python runtime is doing: it's closing the file handle at collection time, which may be too late. Let's say we've missed one scenario in which we need to call `dispose()` explicitly, then many undisposed, garbage keys may have accumulated, which is still a leak. I would add a warning here so that we're alerted if we missed a situation in which an explicit call to `dispose()` is needed. Something along the lines of:

```suggestion
    def dispose(self, *, garbage=False):
        if garbage and not self.disposed:
            warnings.warn("S3 key was not disposed of in time", ResourceWarning)
        if self.multipart is not None:
            for part in self.multipart.parts.values():
                part.dispose(garbage=garbage)
        if self._value_buffer:
            self._value_buffer.close()
        self.disposed = True

    def __del__(self):
        self.dispose(garbage=True)

```

I know this is a high bar to pass, what do you think?

Reply:
Having a warning here is indeed useful, to notify us that we've missed a scenario. Implementing that does expose some techdebt in our code, unfortunately; it looks like we open FileHandles everywhere without closing them properly. TBC! I'll continue to see if there is a cleaner solution.

Reply:
As far as I can tell, all scenarios are now covered. I'll close this, but let me know if I've missed anything. Thanks again for raising this and reviewing this @hannes-ucsc!

Reply:
https://github.com/spulec/moto/issues/5584